### ISSUE TITLE
Laravel keep using html components in markdown mailable.

### ISSUE THREAD

**Issue by przemoli** [laravel/framework#20791]
- Laravel Version: 5.4.28 (and 5.4.33 upgraded in copy of code base)
- PHP Version: 7.0
- Database Driver & Version:

### Description:
`->markdown()` renders main template in markdown, however each `@component` is rendered accordingly to html template (and markdown there do not work - so markdown compilation pass for main template happens before components are in-lined).

### Steps To Reproduce:
Create email view that will use `@component`.

Either:

    php artisan make:mail --markdown=aaa

Or

    $this->view('template')->with(['a' => $a]) into $this->markdown('template', ['a' => $a])

In both cases I was editing templates that come standard after php artisan vendor:publish. Only changes to templates in `html` folder where reflected.

**Comment by przemoli** [comment]
Update on my investigations:

**Unit tests do not test use of `@component` in Markdown Mailable.**

`Mailable` for `->markdown` will eventually call `buildMakdownView` which will return return array of `html` and `text`. `html` is created with `$markdown->render($this->markdown, $data)`, but `render` **will not** use markdown components! Instead it uses components in `html/` folder. _I think that this is root cause of my troubles_. PHPDoc for `render` says " Render the Markdown template into HTML."


I will check if changing which components `render` use, would fix this issue.
If that's it I will try to prepare proper test case too.

**Comment by przemoli** [comment]
From the docs:

* Markdown mails can use Blade components
* Blade components are html only
* Blade components can explicitly compile markdown (hence exaple of markdown table as blade component)

From tests:
* Mailable is meant to set `htmlComponentPath` as location of components.

Why then `Mail\resrouces\views` contain both `html` and `markdown` folders with component definitions? Clearly `markdown` components will never be used.

**Comment by thecrypticace** [comment]
It is my understanding that HTML emails will render using the html-specific components and plain text emails will render with the markdown-specific components.

As I understand it, emails sent by Laravel utilizing the markdown system automatically generate both versions and send them in a single message. Your email client will pick the appropriate one to display.

**Comment by matzeso** [comment]
I am experiencing the same issue. @thecrypticace the problem is not that both versions are sent to the client and the email client will pick one, but the problem is that nested components utilizing markdown are not parsed correctly.

 have a mail::message component and within that I use a mail::button component (a standard layout as can be found within the docs, see [here](https://laravel.com/docs/5.5/mail#writing-markdown-messages))
But the result in e.g. Gmail is that the inner component (the button) is rendered as HTML (I guess first?) and then the outer component is rendered as markdown. As a result, instead of a button in my E-Mail I see the plain HTML code.

This shouldn't be closed as I think this might be a bug.

**Comment by lscarneiro** [comment]
@matzeso I had the same exact problem when opening a edited markdown.

Came to discover that the problem, in my case ofc, was purely due to indentation!
I love to call auto-format in my PHPStorm IDE, and when I did it on this markdown template, it shifted my block from this:
```
...somecode...
@component('mail::button', ['url' => $url])
    button text
@endcomponent
...somecode...
```
to this:
```
...somecode...
    @component('mail::button', ['url' => $url])
        button text
    @endcomponent
...somecode...
```
and then my hell started, how I found the solution? When you are deep in the hell you start to try hopeless solutions, hahaha, that one fixed my bug though!

**Comment by tedsecretsource** [comment]
@thecrypticace I honestly think @przemoli is on to something as I am having a similar problem, but I don't think he/she did a great job of explaining the problem.

*Summary*

Modifications to published markdown email view files, components in particular, are not rendering when previewed in the browser.

*Preconditions*
* All markdown and html email view files have been published to resources/views/vendor/mail [as shown in the documentation](https://laravel.com/docs/5.5/mail#customizing-the-components)
* A mailable has been generated [according to the documentation](https://laravel.com/docs/5.5/mail#generating-mailables) and the `markdown()` method is being used when calling the view
* A route has been defined [according to the documentation](https://laravel.com/docs/5.5/mail#previewing-mailables-in-the-browser) for previewing your work
* We're modifying the files in resources/views/vendor/mail/markdown (the files inside the markdown folder - message.blade.php in particular)

*Steps to Reproduce*
1. Modify resources/views/vendor/mail/markdown/message.blade.php by changing this:
``` php
    {{-- Footer --}}
    @slot('footer')
        @component('mail::footer')
            © {{ date('Y') }} {{ config('app.name') }}. All rights reserved.
        @endcomponent
    @endslot
```
to this
``` php
    {{-- Footer --}}
    @slot('footer')
        @component('mail::footer')
            This text will not render unless I modify the HTML version of this file and put it there even though I'm calling markdown() and indeed my mail blade file seems to be getting processed as markdown.
        @endcomponent
    @endslot
```
2. Preview the formatted email via the route defined in web.php (as per the instructions, in my case "/mailable").

*Expected Result*
* The footer of the email contains "This text will not render unless I modify the HTML version of this file and put it there even though I'm calling markdown() and indeed my mail blade file seems to be getting processed as markdown."

*Actual Result*
* The cited text doesn't appear in the footer. Rather, the default copyright text appears.

*Workaround*
* Modify the HTML file's footer instead

Please consider re-opening. This behavior is reproducible and indentation does not seem to be a factor in this scenario as suggested by @lscarneiro

Thanks in advance.

Ted Stresen-Reuter

**Comment by tedsecretsource** [comment]
Yes, I can confirm that something very strange is happening. I can confirm the following:

* All changes below are in views/vendor/mail/html/message.blade.php
* Only the files in the html folder are being called
* The following is processed as markdown and converted to HTML (expected behavior? what is the purpose of the markdown folder if files in html are being processed as markdown?):
``` php
        @component('mail::footer')
[test](https://sldfld.com)
        @endcomponent
```
* The following is passed through with no markdown processing (that I can see), expected behavior?:
``` php
        @component('mail::footer')
            <a href="https://slfkjs.com">Test</a>
        @endcomponent
```
* The following renders the first part as HTML and the second part as text (HTML is converted to entities - "with bold" is rendered as `&lt;b&gt;with bold&lt;/b&gt;` rather than as **with bold**):
``` php
        @component('mail::footer')
            <div>Test</div>
            This is some more text <b>with bold</b>
        @endcomponent
```

So, in short, it looks like the markdown() function is trying to process files as markdown and trying to be intelligent about it (allowing for normal HTML - which I think is supported) but the question is, why are the two published folders called "html" and "markdown" if only the "html" folder is actually used?

Hope that helps.

**Comment by przemoli** [comment]
To clarify my problem:

I thought markdown is supported in email templates. **IT'S NOT** Blaze templating engine do not support markdown. And if any template do support markdown notation, that's because template contain, PHP snippet that explicitly convert markdown to html.

**Comment by amadeann** [comment]
I think I finally figured this out:

- HTML views are used to render HTML version of markdown emails
- Markdown views are used for the **plain text** version of markdown emails

You can test it by opening the message in some email client which allows you to switch between HTML and plain text view (e.g. [mailtrap.io](https://mailtrap.io/)).

**Comment by przemoli** [comment]
@amadeann 
Yes, you are correct.
However, HTML version use Blade engine, and it allows php snippets, so you can parse some markdown, and I think at least one stock "HTML" component does exactly that, which adds to the confusion.

Also I realized that I had SO question, with accepted answer. So here is the link:
https://stackoverflow.com/questions/45828322/laravel-5-4-wrongly-mix-html-components-in-markdown-mailable

**Comment by tedsecretsource** [comment]
Thanks @przemoli for the SO link. However, the solution really just confirms the slightly confusing nature of the current implementation. Keeping this on my back burner until I have time to fully explore and submit a proper solution.

**Comment by StefanNeuser** [comment]
> I think I finally figured this out:
> 
> * HTML views are used to render HTML version of markdown emails
> * Markdown views are used for the **plain text** version of markdown emails
> 
> You can test it by opening the message in some email client which allows you to switch between HTML and plain text view (e.g. [mailtrap.io](https://mailtrap.io/)).

i think that not correct, what i found out is, that the markdown gets also rendered as HTML. Why else you can use in MailMessage $this->mardown and $this->view?

**Comment by irayoun** [comment]
You can just set all the line codes in the left zero, such as:

`
@component('mail::message')
# Order Shipped

Your order has been shipped!

@component('mail::button', ['url' => $url])
View Order
@endcomponent

Thanks,<br>
{{ config('app.name') }}
@endcomponent
 `
See [https://laravel.com/docs/master/mail#writing-markdown-messages](https://laravel.com/docs/master/mail#writing-markdown-messages)

**Comment by tuncaybahadir** [comment]
> 
> 
> You can just set all the line codes in the left zero, such as:
> 
> ```
> @component('mail::message')
> # Order Shipped
> 
> Your order has been shipped!
> 
> @component('mail::button', ['url' => $url])
> View Order
> @endcomponent
> 
> Thanks,<br>
> {{ config('app.name') }}
> @endcomponent
> ```
> 
> See https://laravel.com/docs/master/mail#writing-markdown-messages

I have been dealing with this problem for 3 days, I solved it as you wrote.

**Comment by hazem-taha** [comment]
m still getting same issue, email is rendered correctly in mailtrap, but not in mail clients like gmail, yahoo,...etc
any chance to fix this ??
